vcd/resource_vcd_independent_disk.go

```diff
@@ -173,33 +197,180 @@ func resourceVcdIndependentDiskCreate(d *schema.ResourceData, meta interface{})
 
 	diskCreateParams.Disk.Description = d.Get("description").(string)
 
+	if value, ok := d.GetOk("sharing_type"); ok {
+		diskCreateParams.Disk.SharingType = value.(string)
+	}
+
 	task, err := vdc.CreateDisk(diskCreateParams)
 	if err != nil {
-		return fmt.Errorf("error creating independent disk: %s", err)
+		return diag.Errorf("error creating independent disk: %s", err)
 	}
 
 	err = task.WaitTaskCompletion()
 	if err != nil {
-		return fmt.Errorf("error waiting to finish creation of independent disk: %s", err)
+		return diag.Errorf("error waiting to finish creation of independent disk: %s", err)
 	}
 
 	diskHref := task.Task.Owner.HREF
 	disk, err := vdc.GetDiskByHref(diskHref)
 	if err != nil {
-		return fmt.Errorf("unable to find disk with href %s: %s", diskHref, err)
+		return diag.Errorf("unable to find disk with href %s: %s", diskHref, err)
 	}
 
 	d.SetId(disk.Disk.Id)
 
-	return resourceVcdIndependentDiskRead(d, meta)
+	return resourceVcdIndependentDiskRead(ctx, d, meta)
+}
+
+func resourceVcdIndependentDiskUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	vcdClient := meta.(*VCDClient)
+
+	if d.HasChanges("size_in_mb", "storage_profile", "description") {
+		_, vdc, err := vcdClient.GetOrgAndVdcFromResource(d)
+		if err != nil {
+			return diag.Errorf(errorRetrievingOrgAndVdc, err)
+		}
+
+		storageProfileValue := d.Get("storage_profile").(string)
+		var storageProfileRef *types.Reference
+
+		if storageProfileValue != "" {
+			storageReference, err := vdc.FindStorageProfileReference(storageProfileValue)
+			if err != nil {
+				return diag.Errorf("error finding storage profile %s", storageProfileValue)
+			}
+			storageProfileRef = &types.Reference{HREF: storageReference.HREF}
+		}
+
+		disk, err := vdc.GetDiskById(d.Id(), true)
+		if err != nil {
+			return diag.Errorf("error fetching independent disk: %s", err)
+		}
+
+		sliceOfVmsHrefs, err := disk.GetAttachedVmsHrefs()
+		if err != nil {
+			return diag.Errorf("[ERROR] resourceVcdIndependentDiskUpdate faced issue fetching attached VMs")
+		}
+
+		//lock VMs as another independent disk resource can be doing update with same VM
+		lockVms(sliceOfVmsHrefs)
+		defer unlockVms(sliceOfVmsHrefs)
+
+		diskDetailsForReAttach, diagErr := detachVms(vcdClient, disk, sliceOfVmsHrefs)
```

Comment:
Should we validate if any of the VMs are online before doing this?

Also a bigger question - should we consistently allow `allow_vm_reboot` like we have it in [internal disk](https://registry.terraform.io/providers/vmware/vcd/latest/docs/resources/vm_internal_disk)?

IIUC, if this fails in the middle - detaches disk from 5 VMs, but the 6th is powered on - the resource will fail and leave it in broken state (5VMs will stay without attached disks)

cc: @lvirbalas , @dataclouder

Reply:
I added an additional check for `IDE` disk type cases to make update atomic.
Also, I corrected the documentation a little. Power off requirement is only for IDE disks, likewise internal disks. So `allow_vm_reboot` is only valid for `IDE` disks, which I guess no one uses :)